stan/math/opencl/kernel_generator/broadcast.hpp

```diff
@@ -0,0 +1,172 @@
+#ifndef STAN_MATH_OPENCL_KERNEL_GENERATOR_BROADCAST_HPP
+#define STAN_MATH_OPENCL_KERNEL_GENERATOR_BROADCAST_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/opencl/matrix_cl_view.hpp>
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/opencl/kernel_generator/type_str.hpp>
+#include <stan/math/opencl/kernel_generator/name_generator.hpp>
+#include <stan/math/opencl/kernel_generator/operation_cl.hpp>
+#include <stan/math/opencl/kernel_generator/as_operation_cl.hpp>
+#include <stan/math/opencl/kernel_generator/is_valid_expression.hpp>
+#include <limits>
+#include <string>
+#include <type_traits>
+#include <set>
+#include <utility>
+
+namespace stan {
+namespace math {
+
+/**
+ * Represents a broadcasting operation in kernel generator expressions.
+ * @tparam T type of arguments
+ * @tparam Colwise whether to broadcast colwise
+ * @tparam Rowwise whether to broadcast rowwise
+ */
+template <typename T, bool Colwise, bool Rowwise>
+class broadcast_
+    : public operation_cl<broadcast_<T, Colwise, Rowwise>,
+                          typename std::remove_reference_t<T>::Scalar, T> {
+ public:
+  using Scalar = typename std::remove_reference_t<T>::Scalar;
+  using base = operation_cl<broadcast_<T, Colwise, Rowwise>, Scalar, T>;
+  using base::var_name;
+
+  /**
+   * Constructor
+   * @param a expression
+   */
+  explicit broadcast_(T&& a) : base(std::forward<T>(a)) {
+    const char* function = "broadcast";
+    if (Colwise) {
+      check_size_match(function, "Rows of ", "a", a.rows(), "", "", 1);
+    }
+    if (Rowwise) {
+      check_size_match(function, "Columns of ", "a", a.cols(), "", "", 1);
+    }
+  }
+
+  /**
+   * Creates a deep copy of this expression.
+   * @return copy of \c *this
+   */
+  inline auto deep_copy() {
+    auto&& arg_copy = this->template get_arg<0>().deep_copy();
+    return broadcast_<std::remove_reference_t<decltype(arg_copy)>, Colwise,
+                      Rowwise>{std::move(arg_copy)};
+  }
+
+  /**
+   * Generates kernel code for this and nested expressions.
+   * @param[in,out] generated set of already generated operations
+   * @param ng name generator for this kernel
+   * @param i row index variable name
+   * @param j column index variable name
+   * @return part of kernel with code for this and nested expressions
+   */
+  inline kernel_parts generate(const std::string& i, const std::string& j,
+                               const std::string& var_name_arg) const {
```

Comment:
[seperate pr]

Reading this over wrt `modify_argument_indices` it's kind of confusing to read `i` and `j` here as the names of the index but in `modify_argument_indices` as the actual index values. It may be good here to change the argument names to something like `i_name` and `j_name`. Or maybe like `row_idx_name` and `col_idx_name`